casper/src/main/resources/PoS.rhox

```diff
@@ -72,35 +117,165 @@ in {
+                new posBalanceCh, totalActiveBondCh, totalBondCh, totalWithdrawCh, committedRewardsCh in {
+                  @posVault!("balance", *posBalanceCh) |
+                  calTotalBonds!(*totalBondCh) |
+                  calTotalActiveBonds!(*totalActiveBondCh) |
+                  calTotalWithdraw!(*totalWithdrawCh) |
+                  calTotalCommittedRewards!(*committedRewardsCh) |
+                  for (@posBalance <- posBalanceCh; @totalBond <- totalBondCh; @activeBonds <- totalActiveBondCh; 
+                       @totalWithdraw <- totalWithdrawCh; @totalCommittedRewards <- committedRewardsCh){
+                    for (@state <<- stateCh) {
+                      retCh!((posBalance , totalBond , activeBonds, totalWithdraw, totalCommittedRewards, state.get("allBonds"), state.get("activeValidators")))
+                    }
+                  }
+                }
+              } |
+              contract calTotalActiveBonds(returnCh) = {
+                new computeSum, computeDelta in {
+                  for (@state <<- stateCh){
+                    @ListOps!("fold", state.get("allBonds").toList(), 0, *computeSum, *returnCh) |
+                    contract computeSum(@(pk, bondAmount), @acc, resultCh) = {
+                      if (state.get("activeValidators").contains(pk)){
+                        resultCh!(acc + bondAmount)
+                      } else {
+                        resultCh!(acc)
+                      }
+                    }
+                  }
+                }
+              } |
+              contract calTotalBonds(returnCh) = {
+                new computeSum, computeDelta in {
+                  for (@state <<- stateCh){
+                    @ListOps!("fold", state.get("allBonds").toList(), 0, *computeSum, *returnCh) |
+                    contract computeSum(@(pk, bondAmount), @acc, resultCh) = {
+                      resultCh!(acc + bondAmount)
+                    }
+                  }
+                }
+              } |
+              contract calTotalWithdraw(returnCh) = {
+                new computeSum in {
+                  for (@state <<- stateCh){
+                    @ListOps!("fold", state.get("withdrawers").toList(), 0, *computeSum, *returnCh) |
+                    contract computeSum(@(pk, (previousBonding, quantinue)), @acc, resultCh) = {
+                      resultCh!(acc + previousBonding)
+                    }
+                  }
+                }
+              } |
+              contract calTotalCommittedRewards(returnCh) = {
+                new computeSum, computeDelta in {
+                  for (@state <<- stateCh){
+                    @ListOps!("fold", state.get("comittedRewards").toList(), 0, *computeSum, *returnCh) |
+                    contract computeSum(@(pk, reward), @acc, resultCh) = {
+                      resultCh!(acc + reward)
+                    }
+                  }
+                }
+              } |
```

Comment:
These contracts are very similar, it would be nice if we they can be extracted to common one.